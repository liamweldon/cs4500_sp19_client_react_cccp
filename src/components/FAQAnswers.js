import React from "react";
import FAQAnswerService from "../services/FAQAnswerService";
import FAQService from "../services/FAQService";
import { Link } from "react-router-dom";
import "./FAQs.scss";
import "./table.scss";

class FAQAnswers extends React.Component {
  constructor(props) {
    super(props);
    this.faqAnswerService = FAQAnswerService.getInstance();
    this.faqService = FAQService.getInstance();
    this.state = {
      faqAnswers: [],
      faqs: [],
      newAnswer: "",
      selectedQuestion: ""
    };
  }
  componentDidMount() {
    this.getFAAs();
    this.getFAQs();
  }

  handleQuestionSelector = evt => {
    this.setState({ selectedQuestion: evt.target.value });
  };

  handleAnswerInput = evt => {
    this.setState({ newAnswer: evt.target.value });
  };

  getFAAs() {
    this.faqAnswerService.findAllFAAs().then(faqAnswers => {
      this.setState({
        faqAnswers: faqAnswers.length ? faqAnswers : []
      });
    });
  }

  getFAQs() {
    this.faqService.findAllFAQs().then(faqs => {
      if (faqs.length) {
        this.setState({
          faqs: faqs
        });
      }
    });
  }

  addAnswer = () => {
    const { selectedQuestion, newAnswer } = this.state;
    this.faqAnswerService.addFAA({answer: newAnswer}).then(res => {
      this.faqService.linkFAAtoFAQ(res.id, selectedQuestion).then(res => {
        this.getFAAs();
      })
    });
  };

  deleteAnswer = id => {
    this.faqAnswerService.deleteFAA(id).then(res => this.getFAAs());
  };

  render() {
    const { faqs, selectedQuestion, newAnswer } = this.state;

    return (
      <div>
        <h3>Frequently Asked Questions Answers</h3>
        <table className="table">
          <tbody>
            <tr className="header-row">
              <td>Question</td>
              <td>Answer</td>
            </tr>
            <tr key={-1}>
              <td>
                <select
                  value={selectedQuestion}
                  className="form-control"
                  onChange={this.handleQuestionSelector}
                >
                  {faqs.map(faq => (
                    <option value={faq.id} key={faq.id}>
                      {faq.title}
                    </option>
                  ))}
                </select>
              </td>
              <td>
                <input value={newAnswer} onChange={this.handleAnswerInput} />
              </td>
              <td>
                <button onClick={this.addAnswer}>
                  <i className="fas fa-plus-square" />
                </button>
              </td>
              <td />
            </tr>
            {this.state.faqAnswers.map(faqAnswer => (
              <tr key={faqAnswer.id}>
                <td>{faqAnswer.question}</td>
                <td>
                  <Link to={`/admin/faas/${faqAnswer.id}`}>
                    {faqAnswer.answer}
                  </Link>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

export default FAQAnswers;
