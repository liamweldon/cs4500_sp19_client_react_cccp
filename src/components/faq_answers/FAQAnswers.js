import React from "react";
import { Link } from "react-router-dom";
import "../FAQs.scss";
import "../table.scss";

const FAQAnswers = ({ faqs, selectedQuestion, newAnswer, eventHandlers }) => {
  return (
    <div>
      <h3>Frequently Asked Questions Answers</h3>
      <table className="table">
        <tbody>
          <tr className="header-row">
            <td>Question</td>
            <td>Answer</td  >
            <td />
          </tr>
          <tr className="faa-row" key={-1}>
            <td>
              <select
                value={selectedQuestion}
                className="form-control"
                onChange={eventHandlers.handleQuestionSelector}
              >
                {faqs.map(faq => (
                  <option value={faq.id} key={faq.id}>
                    {faq.title}
                  </option>
                ))}
              </select>
            </td>
            <td>
              <input
                value={newAnswer}
                onChange={eventHandlers.handleAnswerInput}
              />
            </td>
            <td>
              <button onClick={eventHandlers.addAnswer}>
                <i className="fas fa-plus-square" />
              </button>
            </td>
            <td />
          </tr>
          {faqs.map(faq => {
            return faq.answers.map(faqAnswer => (
              <tr key={faqAnswer.id}>
                <td>{faqAnswer.question}</td>
                <td>
                  <Link to={`/admin/faas/${faqAnswer.id}`}>
                    {faqAnswer.answer}
                  </Link>
                </td>
                <td>
                  <button
                    onClick={() => {
                      eventHandlers.deleteAnswer(faq.id, faqAnswer.id);
                    }}
                  >
                    <i className="fas fa-trash-alt" />
                  </button>
                </td>
              </tr>
            ));
          })}
        </tbody>
      </table>
    </div>
  );
};

export default FAQAnswers;
