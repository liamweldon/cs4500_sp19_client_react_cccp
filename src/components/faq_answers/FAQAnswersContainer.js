import React from "react";
import FAQAnswerService from "../../services/FAQAnswerService";
import FAQService from "../../services/FAQService";
import FAQAnswers from "./FAQAnswers";

class FAQAnswersContainer extends React.Component {
  constructor(props) {
    super(props);
    this.faqAnswerService = FAQAnswerService.getInstance();
    this.faqService = FAQService.getInstance();
    this.state = {
      faqs: [],
      newAnswer: "",
      selectedQuestion: ""
    };
  }
  componentDidMount() {
    this.getFAQs();
  }

  handleQuestionSelector = evt => {
    this.setState({ selectedQuestion: evt.target.value });
  };

  handleAnswerInput = evt => {
    this.setState({ newAnswer: evt.target.value });
  };

  getFAQs = () => {
    this.faqService.findAllFAQs().then(faqs => {
      if (faqs.length) {
        this.setState({
          faqs: faqs,
          selectedQuestion: faqs[0].id
        });
      }
    });
  };

  addAnswer = () => {
    const { selectedQuestion, newAnswer } = this.state;
    this.faqAnswerService
      .addFAA(selectedQuestion, { answer: newAnswer })
      .then(res => {
        this.getFAQs();
        this.setState({ newAnswer: "" });
      });
  };

  deleteAnswer = (questionId, faaId) => {
    this.faqAnswerService
      .deleteFAA(questionId, faaId)
      .then(res => this.getFAQs());
  };

  render() {
    const eventHandlers = {
      deleteAnswer: this.deleteAnswer,
      handleQuestionSelector: this.handleQuestionSelector,
      handleAnswerInput: this.handleAnswerInput,
      addAnswer: this.addAnswer
    };

    return <FAQAnswers {...this.state} eventHandlers={eventHandlers} />;
  }
}

export default FAQAnswersContainer;
