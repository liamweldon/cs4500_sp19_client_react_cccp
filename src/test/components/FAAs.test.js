import React from 'react';
import renderer from 'react-test-renderer';
import faqsJson from '../mockData/faq.mock.json';
import {StaticRouter} from 'react-router';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import FAQAnswersContainer from "../../components/faq_answers/FAQAnswersContainer";
import FAQAnswers from "../../components/faq_answers/FAQAnswers";

Enzyme.configure({adapter: new Adapter()});

test('FAAs Summary Snapshots', () => {
    const component = renderer.create(
        <StaticRouter location="/admin/faas" context={{}}>
            <FAQAnswers faqs={faqsJson} editing={{}} newAnswer="" selectedQuestion="" eventHandlers={{}} />
        </StaticRouter>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('FAAs summary DOM', () => {
    let wrapper = mount(
        <StaticRouter location="/admin/faqs" context={{}}>
            <FAQAnswersContainer />
        </StaticRouter>
    );

    let faasComponent = wrapper.find(FAQAnswersContainer);

    faasComponent.getFAQs = jest.fn().mockReturnValue(faqsJson);
    wrapper.update();
    faasComponent.setState({faqs: faqsJson, selectedQuestion: faqsJson[0].id});

    // there should be a row in the table for each FAQ
    expect(wrapper.find('.faq-row.existing').length).toEqual(0);
    expect(
        wrapper
            .find('.faq-row.existing')
            .at(0)
            .find('input').length
    ).toEqual(0);
});
