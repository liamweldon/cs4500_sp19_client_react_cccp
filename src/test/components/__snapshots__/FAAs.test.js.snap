// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAAs Summary Snapshots 1`] = `
<div>
  <h3>
    Frequently Asked Questions Answers
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr
        className="header-row"
      >
        <td>
          Question
        </td>
        <td>
          Answer
        </td>
        <td />
      </tr>
      <tr
        className="faa-row"
      >
        <td>
          <select
            className="form-control"
            value=""
          >
            <option
              value={172}
            >
              Background Check
            </option>
            <option
              value={192}
            >
              Employees
            </option>
            <option
              value={232}
            >
              Games
            </option>
          </select>
        </td>
        <td>
          <input
            value=""
          />
        </td>
        <td>
          <button>
            <i
              className="fas fa-plus-square"
            />
          </button>
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
